%Constant
\newcommand\NGSLibrairie{SureSelect QXT FH\_Angers\_v1.0}
\newcommand\NGSSequenceur{Illumina NextSeq500 2x150bp}

%Mutations are extracted from the mutations .csv file

%Constant
%Could be extracted from a .json file, as Pierre G. does
\newcommand\BioinfoFoxNGSVersion{1.0.0}
\newcommand\BioinfoGenome{GRCh37 (hg19, 2013)}
\newcommand\BioinfoDemultiplexingTool{blc2fastq}
\newcommand\BioinfoDemultiplexingVersion{2.20.0.422}
\newcommand\BioinfoAlignmentTool{bwa-mem}
\newcommand\BioinfoAlignmentVersion{0.7.17-r1188}
\newcommand\BioinfoVariantCallingATool{HaplotypeCaller}
\newcommand\BioinfoVariantCallingBTool{Mutect2}
\newcommand\BioinfoVariantCallingCTool{Varscan}
\newcommand\BioinfoVariantCallingDTool{Pindel}
\newcommand\BioinfoVariantCallingAVersion{4.2.6.1}
\newcommand\BioinfoVariantCallingBVersion{4.2.6.1}
\newcommand\BioinfoVariantCallingCVersion{2.3.9}
\newcommand\BioinfoVariantCallingDVersion{0.2.5b9}
\newcommand\BioinfoClassificationTool{Annovar}
\newcommand\BioinfoClassificationVersion{2020-06-07}
\newcommand\BioinfoAnnotationList{COSMIC (v.89 \& v.92), ClinVar (2020-03-16), GnomAD (Exome v2.1), avsnp138, IARC}
\newcommand\BioinfoSeuilVAF{2}
\newcommand\BioinfoSeuilAF{1}
\newcommand\BioinfoSeuilPolymorphism{0.1}